addons/web_editor/static/src/js/wysiwyg/widgets/media.js

```diff
@@ -551,36 +493,72 @@ var FileWidget = SearchableMediaWidget.extend({
     /**
      * @private
      */
-    _onAttachmentClick: function (ev, doubleClick) {
+    _onAttachmentClick: function (ev, save) {
         var $attachment = $(ev.currentTarget);
         var attachment = _.find(this.attachments, {id: $attachment.data('id')});
-        this._toggleImage(attachment, doubleClick);
+        this._selectAttachement(attachment, save);
     },
     /**
      * @private
      */
     _onAttachmentDblClick: function (ev) {
         this._onAttachmentClick(ev, true);
-        this.trigger_up('save_request');
     },
     /**
      * @private
+     * @returns {Promise}
      */
     _onFileInputChange: function () {
-        this.$el.addClass('nosave');
-        this.$form.removeClass('o_has_error').find('.form-control, .custom-select').removeClass('is-invalid');
-        this.$errorText.empty();
-        this.$uploadButton.removeClass('btn-danger btn-success');
-        this._uploadFile();
+        return this._mutex.exec(this._addDatas.bind(this));
+    },
+    /**
+     * Uploads the files that are currently selected on the file input, which
+     * creates new attachments. Then inserts them on the media dialog and
+     * selects them. If multiImages is not set, also triggers up the
+     * save_request event to insert the attachment in the DOM.
+     *
+     * @private
+     * @returns {Promise}
+     */
+    _addDatas: function () {
+        var self = this;
+        var uploadMutex = new concurrency.Mutex();
+
+        var files = _.sortBy(this.$fileInput[0].files, 'size');
+
+        _.each(files, function (file) {
+            // upload one file at a time
+            uploadMutex.exec(function () {
+                return utils.getDataURLFromFile(file).then(function (result) {
+                    return self._rpc({
+                        route: '/web_editor/attachment/add_data',
```

Comment:
The idea is that if the user selects 10 big files:
- with the mutex: 1st file uploaded alone (10s) -> open optimize modal for this file, other files upload in background, hopefully 2nd upload is done when optimize of 1st file is done, etc.
- without mutex: 10 files uploaded at the same time (100s) -> delays opening the optimize modal at the end, and then nothing happens in background

Reply:
The optimize modal is opened for each file individually in turn ? Why not uploading at the same time (which is less than 100s in your example since this could be done in parallel) and opening the optimize modal for whichever file is ready first (while the others are still being uploaded) ? That way if you have one big files and 9 small files, the modal will open really fast. I think the code would not be more complex (it is just moving the mutex elsewhere I think ?).

Reply:
Upload is typically limited by bandwidth, it's hard to get parallel in practice, every additional file uploaded at the same time will delay all the others.